File: endpoint-client/models/firewall_rule.py
class FirewallRule:
    """
    Model representing a firewall rule.
    """
    def __init__(self, rule_id=None, name=None, description=None, source_ip=None, 
                 destination_ip=None, source_port=None, destination_port=None, 
                 protocol=None, action=None, priority=0, enabled=True, 
                 application_id=None, created_at=None, updated_at=None):
        self.id = rule_id
        self.name = name
        self.description = description
        self.sourceIp = source_ip
        self.destinationIp = destination_ip
        self.sourcePort = source_port
        self.destinationPort = destination_port
        self.protocol = protocol
        self.action = action
        self.priority = priority
        self.enabled = enabled
        self.applicationId = application_id
        self.createdAt = created_at
        self.updatedAt = updated_at
    
    @classmethod
    def from_dict(cls, data):
        """
        Create a FirewallRule instance from a dictionary.
        """
        return cls(
            rule_id=data.get('_id') or data.get('id'),
            name=data.get('name'),
            description=data.get('description'),
            source_ip=data.get('sourceIp'),
            destination_ip=data.get('destinationIp'),
            source_port=data.get('sourcePort'),
            destination_port=data.get('destinationPort'),
            protocol=data.get('protocol'),
            action=data.get('action'),
            priority=data.get('priority', 0),
            enabled=data.get('enabled', True),
            application_id=data.get('applicationId'),
            created_at=data.get('createdAt'),
            updated_at=data.get('updatedAt')
        )
    
    def to_dict(self):
        """
        Convert the FirewallRule instance to a dictionary.
        """
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'sourceIp': self.sourceIp,
            'destinationIp': self.destinationIp,
            'sourcePort': self.sourcePort,
            'destinationPort': self.destinationPort,
            'protocol': self.protocol,
            'action': self.action,
            'priority': self.priority,
            'enabled': self.enabled,
            'applicationId': self.applicationId,
            'createdAt': self.createdAt,
            'updatedAt': self.updatedAt
        }
    
    def __repr__(self):
        return f"<FirewallRule {self.name}>"

File: mongodb-backend/src/models/FirewallRule.js
const mongoose = require('mongoose');

const firewallRuleSchema = new mongoose.Schema({
  endpointId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Endpoint',
    default: null  // Optional - null for frontend rules
  },
  applicationId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Application',
    default: null  // Optional - can create standalone rules
  },
  processName: {
    type: String,
    trim: true,
    default: null  // Optional - only for NGFW rules
  },
  name: {
    type: String,
    trim: true
  },
  description: {
    type: String,
    trim: true
  },
  // NGFW flow fields
  entity_type: {
    type: String,
    enum: ['ip', 'domain'],
    default: 'ip'  // Default to 'ip' instead of required
  },
  // For IP-based rules
  source_ip: {
    type: String,
    trim: true
  },
  destination_ip: {
    type: String,
    trim: true
  },
  source_port: {
    type: Number
  },
  destination_port: {
    type: Number
  },
  // For domain-based rules
  domain_name: {
    type: String,
    trim: true
  },
  action: {
    type: String,
    enum: ['allow', 'deny', 'ALLOW', 'DENY'],
    required: [true, 'Action is required']
  },
  // Legacy fields for backward compatibility
  entityType: {
    type: String,
    enum: ['ip', 'domain'],
    default: 'ip'
  },
  domain: {
    type: String,
    trim: true
  },
  sourceIp: {
    type: String
  },
  destinationIp: {
    type: String
  },
  sourcePort: {
    type: Number
  },
  destinationPort: {
    type: Number
  },
  protocol: {
    type: String,
    enum: ['TCP', 'UDP', 'ICMP', 'ANY'],
    default: 'ANY'
  },
  priority: {
    type: Number,
    default: 0
  },
  enabled: {
    type: Boolean,
    default: true
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
}, {
  timestamps: true
});

// Index for faster queries
firewallRuleSchema.index({ endpointId: 1 });
firewallRuleSchema.index({ processName: 1 });
firewallRuleSchema.index({ endpointId: 1, processName: 1 });
firewallRuleSchema.index({ applicationId: 1 });
firewallRuleSchema.index({ priority: 1 });
firewallRuleSchema.index({ enabled: 1 });

const FirewallRule = mongoose.model('FirewallRule', firewallRuleSchema);

module.exports = FirewallRule;
